### PR TITLE
Enable using dash or underscore in CLI

### PR DESCRIPTION
**Pull request by muellerzr**
# What does this PR do?

This PR adds a pinch of magic for users that are used to doing `--a-b` with arguments that pass-through to things such as `torchrun`, accelerate can process those if we have them defined ourselves already.

E.g. a user can do:
```bash
accelerate launch --no_python --multi_gpu --num_processes 8 --monitor_interval=1 ./trampoline.sh python3 -c "print('hello')"
```
**OR**:
```bash
accelerate launch --no-python --multi-gpu --num-processes 8 --monitor-interval=1 ./trampoline.sh python3 -c "print('hello')"
```

Fixes https://github.com/huggingface/accelerate/issues/2241

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/accelerate/blob/main/CONTRIBUTING.md#submitting-a-pull-request-pr),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/accelerate/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/accelerate/tree/main/docs#writing-documentation---specification).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@BenjaminBossan, @stas00 (for your future sanity 😄 )

### CODE REVIEW

**Comment by stas00** [comment]
Zach, in the OP, the 2 lines of code are identical - a copy-n-paste mistake?

**Comment by muellerzr** [comment]
OP was posted before I did this 😅 I'll go update it!

**Comment by muellerzr** [comment]
Actually, was weird phrasing on my end. Edited it to be more succinct :)

**Comment by stas00** [comment]
I meant this:
```
accelerate launch --no_python --multi_gpu --num_processes 8 --monitor_interval=1 ./trampoline.sh python3 -c "print('hello')"
accelerate launch --no-python --multi-gpu --num-processes 8 --monitor-interval=1 ./trampoline.sh python3 -c "print('hello')"
```

**Comment by stas00** [comment]
oh, wait, they aren't identical ;) it's the `multi_gpu` vs `multi-gpu`

File: src/accelerate/commands/utils.py [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+        if len(option_tuples) > 1:
+            options = ", ".join([option_string for _, option_string, _ in option_tuples])
+            self.error(f"ambiguous option: {arg_string} could match {options}")
+
+        # If exactly one match, return it
+        elif len(option_tuples) == 1:
+            (option_tuple,) = option_tuples
+            return option_tuple
+
+        # If not found, but looks like a negative number, probably posisional
```
Comment (BenjaminBossan) [inline src/accelerate/commands/utils.py:45]:
```suggestion
        # If not found, but looks like a negative number, probably positional
```

File: tests/test_cli.py [hunk 2]
```diff
@@ -99,6 +100,50 @@ def test_notebook_launcher(self):
             run_command(cmd, env=os.environ.copy())
 
 
+class LaunchArgTester(unittest.TestCase):
+    """
+    Test cases revolving around the CLI wrappers
+    """
+
+    def test_hiphen(self):
```
Comment (BenjaminBossan) [inline tests/test_cli.py:108]:
```suggestion
    def test_hyphen(self):
```

File: src/accelerate/commands/launch.py [hunk 3]
```diff
@@ -127,10 +128,13 @@ def __call__(self, parser, namespace, values, option_string=None):
 
 
 def launch_command_parser(subparsers=None):
+    description = "Launch a python script in a distributed scenario. Arguments can be passed in with either hyphens (`a-b`) or underscores (`a_b`)"
```
Comment (BenjaminBossan) [inline src/accelerate/commands/launch.py:131]: I think the doc would be clearer with a real example, for `num-processes` instead of `a-b`.

**Review by BenjaminBossan — COMMENTED** [review]
I'm +/- in this PR. I understand the goal to enable both ways of writing this, but I'm uncertain if the extra complexity is worth it.

We override two private methods, which in turn use more private methods. Are we sure that they work in all edge cases with all Python versions? Is the new code 100% covered by tests? If there is an issue there, it has the potential to break our users' scripts.

I went looking for alternatives. `argparse` has the concept of custom `Action`s, which could be a solution, but the documentation wasn't very helpful IMO:

https://docs.python.org/3/library/argparse.html#action-classes

Also, at a different place, it says "New in version 3.9", so not sure if this works with 3.8.

Maybe another way would be to detect if there are missing arguments with `_` in the name that, if replaced by `-`, would match an existing argument, and then add a useful warning, but leave the arguments untouched otherwise? This could be less disruptive.

Also note:
- [click](https://github.com/pallets/click/issues/1123) made the change to only use `-`
- [typer](https://github.com/tiangolo/typer/issues/341) automatically uses `-` instead of `_`

File: src/accelerate/commands/utils.py [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import argparse
+
+
+class ArgumentParserWithDashSupport(argparse.ArgumentParser):
```
[the remark this answers is not in the document]
Comment (muellerzr) [reply src/accelerate/commands/utils.py:17]: The aim is to not bloat `--help`, these are passthrough arguments to downstream API hook-ins such as `torchrun` which uses `-` rather than `_`.
